### PR TITLE
SAML SP-initiated and IdP-initiated flows

### PR DESCRIPTION
**Pull request by chanioxaris**
Introduce a new page that explains how the two different authentication flows, SP-initiated and IdP-initiated, work and their differences

### CODE REVIEW

File: docs/authentication/saml/authentication-flows.mdx [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
+title: Authentication flows
+description: Learn about the SAML SSO Authentication flows.
+---
+
+# Authentication flows
+
+The SAML protocol supports two different methods in order to start an SSO flow: SP-initiated and IdP-initiated. Below we will explore
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:8]:
```suggestion
The SAML protocol supports two different methods in order to start an SSO flow: SP-initiated and IdP-initiated. This guide will explore
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+how each of them works and the differences between the two.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:9]:
```suggestion
the differences between the two.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+## SP-initiated flow
+In an SP-initiated flow, the user is starting the authentication flow from your application (SP), by providing the email address. Then they are being redirected to the SAML provider (IdP) where
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:12]:
```suggestion
In an SP-initiated (service-provider-initiated) flow:

- The user starts the authentication flow from your application by providing the email address.
- The user is redirected to the SAML provider (IdP), where they must authenticate themselves.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+they need to authenticate themselves. After successful authentication, the user is finally redirected back to your application, gaining access to their account.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:13]:
```suggestion
- After successful authentication, the user is redirected back to your application, gaining access to their account.
```

Comment (S3Prototype) [reply docs/authentication/saml/authentication-flows.mdx:13]: We should mention what happens when authentication fails, for completeness

File: docs/authentication/saml/authentication-flows.mdx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+## IdP-initiated flow
+In a IdP-initiated flow, the user is starting the authentication flow from the SAML provider (IdP), by selecting which application (SP) they would like to access.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:16]:
```suggestion
In an IdP-initiated (Identity-Provider-initiated) flow:

- The user starts the authentication flow from the SAML provider (IdP) by selecting which application (or service provider) they would like to access.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+Then they are being redirected to the application of their choice, gaining access to their account.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:17]:
```suggestion
The user is redirected to the application of their choice, gaining access to their account.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+<Callout type="info">
+    IdP-Initiated flow carries a security risk. We recommend using a SP-Initiated flow whenever possible.
+</Callout>
+
+To allow IdP-initiated flows for your Enterprise Connection:
+
+1. Navigate to the [**Enterprise Connections**](https://dashboard.clerk.com/last-active?path=user-authentication/enterprise-connections) page in the Clerk Dashboard and click on the desired connection.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:25]:
```suggestion
1. Navigate to the [**Enterprise Connections**](https://dashboard.clerk.com/last-active?path=user-authentication/enterprise-connections) page in the Clerk Dashboard and select the desired connection.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+2. Select the **Advanced** tab at the top.
+
+3. Enable by checking the **Allow IdP-Initiated flow** toggle.
+
+### Risks of IdP-initiated flow
+By enabling IdP-Initiated flow, you accept Clerk to receive unsolicited authentication requests by the Identity Provider. That means that neither party, SP or IdP, can verify that the flow has been initiated by this specific user.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:32]:
```suggestion
By enabling IdP-Initiated flow, you allow Clerk to receive unsolicited authentication requests by the Identity Provider. Neither the Service Provider or the Identity Provider will be able to verify that the flow was initiated by the specified user.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+As a result of the above, the flow is susceptible to MITM (man-in-the-middle) attacks where a bad actor can steal the IdP response and use it to gain access to a compromised account.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:34]:
```suggestion
As a result, the flow is susceptible to MITM (man-in-the-middle) attacks in which a bad actor hijacks the IdP response and uses it to gain access to a compromised account.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+Also, a bad actor can intercept the IdP response and replace it with another, in order to make the target user to sign in as the attacker, performing a Login CSRF attack.
+
+### Mitigation of IdP-initiated flow risks
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:38]:
```suggestion
```
Don't need this

File: docs/authentication/saml/authentication-flows.mdx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+Even if the IdP-initiated is considered an insecure flow, Clerk tries to mitigate those flaws in order to protect and offer the best possible security for your application and users.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:39]:
```suggestion
Even if the IdP-initiated flow is considered an insecure, Clerk tries to mitigate those flaws in order to protect and offer the best possible security for your application and users.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+#### Unsolicited InResponseTo attribute
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:41]:
```suggestion
#### Unsolicited `InResponseTo` attribute
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+Based on the [SAML 2.0 profiles specification](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf), we are making sure that an unsolicited response does not contain an `InResponseTo` attribute.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:42]:
```suggestion
In accordance with the [SAML 2.0 profiles specification](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf), Clerk ensures that an unsolicited response does not contain an `InResponseTo` attribute.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+This prevents bad actors of stealing a response used in a SP-initiated flow and then try to use it in an IdP-initiated flow.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:43]:
```suggestion
This prevents bad actors from stealing a response used in an SP-initiated flow and using it in an IdP-initiated flow.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+#### Replay detection
+We prevent responses from being re-used, by consuming them and remembering which have already been used.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:46]:
```suggestion
We prevent responses from being re-used by consuming them and remembering which have already been used.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+This prevents bad actors of stealing a response and then trying to re-use it in order to get access.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:47]:
```suggestion
This prevents bad actors from stealing and reusing a response to gain access to a user's account.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+#### Use small validation periods
+Each SAML response defines when they were issued and when they are expiring. As an IdP-initiated flow is expected to be completed
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:50]:
```suggestion
Each SAML response defines when they were issued and when they will expire. As an IdP-initiated flow is expected to be completed
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+within a timeframe of seconds, you need to make sure that these validation periods is as small as possible.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:51]:
```suggestion
within a timeframe of seconds, you must ensure that these validation periods are as small as possible.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+#### Multi-Factor
+As it's happening with every other flow, MFA is also applied to the SAML IdP-initiated flow with no exception. If user has set up a MFA method to their account,
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:54]:
```suggestion
MFA is available for the SAML IdP-initiated flow.
```

File: docs/authentication/saml/authentication-flows.mdx [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+then they will need to provide and verify it, before gaining any access to the application.
```
Comment (S3Prototype) [inline docs/authentication/saml/authentication-flows.mdx:55]:
```suggestion
```

**Review by S3Prototype — APPROVED** [review]
Nice! Approved with suggestions

File: docs/authentication/saml/authentication-flows.mdx (hunk 3 above)
Comment (chanioxaris) [reply docs/authentication/saml/authentication-flows.mdx]: As i am aware the term that SAML is using is `SP-initiated` all the time, so it guess it's fine to leave it as is?

File: docs/authentication/saml/authentication-flows.mdx (hunk 4 above)
Comment (chanioxaris) [reply docs/authentication/saml/authentication-flows.mdx]: Nothing fancy and it depends on the SAML provider. I guess the most common scenario is that the provider just returns an error on their sign in page. I guess it's fine to focus only on the happy path here?

Comment (S3Prototype) [reply docs/authentication/saml/authentication-flows.mdx]: Ah, I see. That's fine then yeah

**Pull request merged**